tools/isobuild/import-scanner.js

```diff
@@ -0,0 +1,431 @@
+    this.absPathToOutputIndex = {};
+    this.outputFiles = [];
+    this.extensions = extensions;
+  }
+
+  addInputFiles(...args) {
+    const Ap = Array.prototype;
+    const flatFiles = Ap.concat.apply(Ap, args);
+
+    flatFiles.forEach(file => {
+      const absPath = pathJoin(this.sourceRoot, file.sourcePath);
+
+      // Files that are not eagerly evaluated (lazy) will only be included
+      // in the bundle if they are actually imported.
+      file.lazy = this._isFileLazy(file);
+
+      // Files that are eagerly evaluated are effectively "imported" as
+      // entry points.
+      file.imported = ! file.lazy;
+
+      file.installPath = this._getInstallInfo(absPath).path;
+
+      if (has(this.absPathToOutputIndex, absPath)) {
+        const index = this.absPathToOutputIndex[absPath];
+        this.outputFiles[index] = file;
+      } else {
+        this.absPathToOutputIndex[absPath] =
+          this.outputFiles.push(file) - 1;
+      }
+    });
+
+    return this;
+  }
+
+  getOutputFiles() {
+    this.outputFiles.forEach(file => {
+      const absPath = pathJoin(this.sourceRoot, file.sourcePath);
+      file.deps = this._scanDeps(absPath, file.data)
+    });
+
+    return this.outputFiles;
+  }
+
+  _isFileLazy(file) {
+    if (has(file, "lazy")) {
+      return !! file.lazy;
+    }
+
+    // If file.lazy was not previously defined, mark the file lazy if it
+    // is contained by an imports directory. Note that any files contained
+    // by a node_modules directory will already have been marked lazy in
+    // PackageSource#_inferFileOptions. The reason we can't do all our
+    // lazy marking in the _inferFileOptions method is that we don't know
+    // then whether the current app or package is using the modules
+    // package. At this point, we know the modules package must be in use,
+    // because the ImportScanner is only ever used when modules are used.
+    return this._splitPath(
+      pathDirname(file.sourcePath)
+    ).indexOf("imports") >= 0;
+  }
+
+  _scanDeps(absPath, data) {
+    const deps = keys(findImportedModuleIdentifiers(data.toString("utf8")));
+
+    each(deps, id => {
+      const absImportedPath = this._tryToResolveImportedPath(id, absPath);
+      if (! absImportedPath) {
+        return;
+      }
+
+      if (has(this.absPathToOutputIndex, absImportedPath)) {
+        // Avoid scanning files that we've scanned before, but mark them
+        // as imported so we know to include them in the bundle if they
+        // are lazy.
+        const index = this.absPathToOutputIndex[absImportedPath];
+        this.outputFiles[index].imported = true;
+        return;
+      }
+
+      const installInfo = this._getInstallInfo(absImportedPath);
+      if (! installInfo) {
+        // The given path cannot be installed on this architecture.
+        return;
+      }
+
+      if (! installInfo.inNodeModules) {
+        // At this point, if the file is not in a node_modules directory,
```

Comment:
> Or maybe what gets passed to addInputFiles in this class is all of the JS files in the package?

Yes, that's what happens now, thanks to this commit: https://github.com/benjamn/meteor/commit/21a0780194bd9b40ad384b80ae66262a010d9174

As @trusktr points out, this may be less efficient, but from what I can tell most packages don't include (m)any files that were never added with `api.addFiles`, so removing those unnecessary calls to `api.addFiles` and adding all files automatically shouldn't be that much more expensive than what we did before.